### PR TITLE
Gas Context for Inter-Contract Calls

### PR DESCRIPTION
**Pull request by miloszm**
This PR implements gas context per stack-frame, making it possible to create local gas limit for every inter-contract call.
The size of gas limit can be either default or given explicitly in the query and transact call as parameter.
"gas_context" test contract has been added to test both query and transact with default or passed limits.
Necessary adjustments have been made in other gas relevant tests as there is no longer a global gas meter.

### CODE REVIEW

**Review by ureeves — CHANGES_REQUESTED** [review]
Looks good overall. Just some changes requested.

For this one I would wait for the review of one other before merging.

File: src/gas.rs
Comment (krl) [inline src/gas.rs:16]: The use of floating point math here could introduce different rounding in subsequent calls to gas_left etc, we should be using fixed-point math for these calculations.

File: src/memory.rs
Comment (krl) [reply src/memory.rs:53]: While i think that is good practice, the unsafe is not introduced in this PR, and should probably be its own issue

File: tests/lib.rs
Comment (krl) [inline tests/lib.rs:456]: Fixed point should also be used here

**Review by krl — CHANGES_REQUESTED** [review]

File: src/call_context.rs
Comment (miloszm) [reply src/call_context.rs:342]: @ureeves your version does not compile

Comment (ureeves) [reply src/call_context.rs:342]: Ah you're right. Immutable borrow while holding a mutable. This could be solved by rearranging things a little. I don't think we should clone the gas meter every time we use `gas_merge`

**Review by krl — APPROVED** [review]

File: src/call_context.rs
Comment (ureeves) [inline src/call_context.rs:26]: Unnecessary lifetime?

File: src/call_context.rs
Comment (ureeves) [inline src/call_context.rs:32]: Unnecessary lifetime?

**Review by ureeves — CHANGES_REQUESTED** [review]
There seem to be a few rogue lifetimes left.

File: src/call_context.rs
Comment (miloszm) [reply src/call_context.rs]: Done

File: src/call_context.rs
Comment (miloszm) [reply src/call_context.rs]: Done

**Review by ureeves — APPROVED** [review]

**Pull request merged**
